src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -11,184 +10,167 @@ namespace System.IO
-                        ok = UnsafeNativeMethods.ReadDirectoryChangesW(directoryHandle,
-                                                           buffPtr,
-                                                           _internalBufferSize,
-                                                           _includeSubdirectories ? 1 : 0,
-                                                           (int)_notifyFilters,
-                                                           out size,
-                                                           overlappedPointer,
-                                                           IntPtr.Zero);
-                    }
-                }
+                // If shutdown has been requested, exit.  The finally block will handle
+                // cleaning up the entire operation, as continueExecuting will remain false.
+                if (!_enabled || IsHandleInvalid(state.DirectoryHandle))
+                    return;
+
+                // Get the overlapped pointer to use for this iteration.
+                overlappedPointer = state.ThreadPoolBinding.AllocateNativeOverlapped(state.PreAllocatedOverlapped);
+                int size;
+                continueExecuting = Interop.mincore.ReadDirectoryChangesW(
+                    state.DirectoryHandle,
+                    state.Buffer, // the buffer is kept pinned for the duration of the sync and async operation by the PreAllocatedOverlapped
+                    _internalBufferSize,
+                    _includeSubdirectories,
+                    (int)_notifyFilters,
+                    out size,
+                    overlappedPointer,
+                    IntPtr.Zero);
             }
             catch (ObjectDisposedException)
-            { //Ignore
+            {
+                // Ignore.  Disposing of the handle is the mechanism by which the FSW communicates
+                // to the asynchronous operation to stop processing.
             }
             catch (ArgumentNullException)
-            { //Ignore
-                Debug.Assert(IsHandleInstanceInvalid(directoryHandle), "ArgumentNullException from something other than SafeHandle?");
+            {
+                //Ignore.  The disposed handle could also manifest as an ArgumentNullException.
+                Debug.Assert(IsHandleInvalid(state.DirectoryHandle), "ArgumentNullException from something other than SafeHandle?");
             }
             finally
             {
-                if (!ok)
+                // At this point the operation has either been initiated and we'll let the callback
+                // handle things from here, or the operation has been stopped or failed, in which case
+                // we need to cleanup because we're no longer executing.
+                if (!continueExecuting)
                 {
-                    threadPoolBinding.FreeNativeOverlapped(overlappedPointer);
+                    // Clean up the overlapped pointer created for this iteration
```

Comment:
Can we make the state var implement IDisposable so we can call state.Dispose and it'll clean up any internal vars correctly without needed to make 3 calls to it?